lightning/src/routing/network_graph.rs

```diff
@@ -594,13 +599,25 @@ impl NetworkGraph {
 		}
 	}
 
-	/// For a new or already known (from previous announcement) channel, store or update channel info.
-	/// Also store nodes (if not stored yet) the channel is between, and make node aware of this channel.
-	/// Checking utxo on-chain is useful if we receive an update for already known channel id,
-	/// which is probably result of a reorg. In that case, we update channel info only if the
-	/// utxo was checked, otherwise stick to the existing update, to prevent DoS risks.
+	/// Store or update channel info from a channel announcement.
+	///
+	/// You probably don't want to call this directly, instead relying on a NetGraphMsgHandler's
+	/// RoutingMessageHandler implementation to call it indirectly. This may be useful to accept
+	/// routing messages without checking their signatures.
+	///
+	/// If the channel has been confirmed to exist on chain (with correctly-formatted scripts on
+	/// chain), set utxo_value to the value of the output on chain, otherwise leave it as None.
+	/// The UTXO value is then used in routing calculation if we have no better information on the
+	/// maximum HTLC value which can be sent over the channel. Further, setting utxo_value to Some
+	/// indicates that the data is genuine, allowing us to update existing data (in the case of
+	/// reorgs).
+	///
 	/// Announcement signatures are checked here only if Secp256k1 object is provided.
-	fn update_channel_from_announcement(&mut self, msg: &msgs::ChannelAnnouncement, utxo_value: Option<u64>, secp_ctx: Option<&Secp256k1<secp256k1::VerifyOnly>>) -> Result<bool, LightningError> {
+	pub fn update_channel_from_announcement<T: secp256k1::Verification>(&mut self, msg: &msgs::ChannelAnnouncement, utxo_value: Option<u64>, secp_ctx: Option<&Secp256k1<T>>) -> Result<bool, LightningError> {
+		if msg.contents.node_id_1 == msg.contents.node_id_2 || msg.contents.bitcoin_key_1 == msg.contents.bitcoin_key_2 {
+			return Err(LightningError{err: "Channel announcement node had a channel with itself".to_owned(), action: ErrorAction::IgnoreError});
+		}
```

Comment:
Does the `handling_channel_announcements` test need to be updated?

Reply:
To test for the node_id/bitcoin_id equality check that got copied to here, or do you mean to test `update_channel_from_announcement` directly?

Reply:
The equality check. Didn't see it was copied from `handle_channel_announcement`. Does the check need to be duplicated there?

Reply:
The check is in `handle_channel_announcement` as a general "we should do simple checks before DoS-vulnerable checks including UTXO lookups" thing. We're a little lazy about it anyway already, however, given we don't check signatures until after UTXO lookups. We can change it if you feel strongly.